sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -744,7 +735,84 @@ public void processElement(@Element UserT element, ProcessContext context)
                               KV.of(hashDestination(destination, destinationCoder), element));
                         }
                       }))
-              .setCoder(KvCoder.of(VarIntCoder.of(), input.getCoder()))
+              .setCoder(KvCoder.of(VarIntCoder.of(), input.getCoder()));
+
+      if (input.getWindowingStrategy().needsMerge()) {
+        // To make sure that the elements are partitioned by original input windows and to work with
+        // merging windows such as sessions, we firstly group the elements by destination to reify
+        // the window, if the window needs to be merged, and then perform sharded writes for each
+        // destination and window pair.
+        Coder<BoundedWindow> windowCoder =
+            (Coder<BoundedWindow>) input.getWindowingStrategy().getWindowFn().windowCoder();
+        return reifyWindowAndWrite(keyedInput, windowCoder, input.getCoder());
+      } else {
+        return write(keyedInput, input.getCoder());
+      }
+    }
+
+    private PCollection<List<FileResult<DestinationT>>> reifyWindowAndWrite(
+        PCollection<KV<Integer, UserT>> keyedInput,
+        Coder<BoundedWindow> windowCoder,
+        Coder<UserT> inputCoder) {
+      PCollection<KV<ValueInSingleWindow<Integer>, UserT>> windowedInput =
+          keyedInput
+              .apply("GroupByDestination", GroupByKey.create())
```

Comment:
This GBK could be a bottleneck although we "reshard" before the write which is potentially more expensive. I don't have a better idea other than reifying the window in the keys.

Reply:
Why do you need to group here? Can't the window be reified without Grouping? Won't groupintobatches perform the shuffle?

I worry that often we may have a single destination, or only a couple of them, so we'd be reducing parallelism to 1.

Reply:
Right that was my concern. We do shuffling for `GroupIntoBatches` but stateful DoFns are not compatible with session windows without an explicit GBK.